README.md
# GEOSwiftMapboxGL

[![Cocoapods Compatible](https://img.shields.io/cocoapods/v/GEOSwiftMapboxGL)](https://cocoapods.org/pods/GEOSwiftMapboxGL)
[![Supported Platforms](https://img.shields.io/cocoapods/p/GEOSwiftMapboxGL)](https://github.com/GEOSwift/GEOSwiftMapboxGL)
[![Build Status](https://img.shields.io/travis/GEOSwift/GEOSwiftMapboxGL/main)](https://travis-ci.com/GEOSwift/GEOSwiftMapboxGL)

See [GEOSwift](https://github.com/GEOSwift/GEOSwift) for full details

## Requirements

* iOS 9.0+
* Swift 5.1+
* CocoaPods 1.4.0+
* MapboxGL 6

## Installation

### CocoaPods

1. Follow the Mapbox Maps SDK [installation
   instructions](https://docs.mapbox.com/ios/maps/guides/install/) to configure
   your `~/.netrc` file.
2. Update your `Podfile` to include:

        use_frameworks!
        pod 'GEOSwiftMapboxGL'

3. Run `$ pod install`

## Usage

```swift

import GEOSwift
import Mapbox
import GEOSwiftMapboxGL

...

let point = Point(longitude: 10, latitude: 45)
let polygon = try! Polygon(wkt: "POLYGON((35 10, 45 45.5, 15 40, 10 20, 35 10),(20 30, 35 35, 30 20, 20 30))")

let annotations = [
    MGLShape.make(with: point),
    MGLShape.make(with: polygon)
]
self.mapView.showAnnotations(annotations, animated: true)

...

```

In this table you can find which annotation class you should expect when using
`MGLShape.make(with:)`:

| GEOSwift | MapboxGL |
|:-------------:|:-----------------:|
| `Point` | `MGLPointAnnotation` |
| `LineString` | `MGLPolyline` |
| `Polygon` | `MGLPolygon` |
| `MultiPoint` | `MGLPointCollection` |
| `MultiLineString` | `MGLMultiPolyline` |
| `MultiPolygon` | `MGLMultiPolygon` |
| `GeometryCollection` | `MGLShapeCollection` |

You can also use `MGLShape.makeFeature(with:)`:

| GEOSwift | MapboxGL |
|:-------------:|:-----------------:|
| `Point` | `MGLPointFeature` |
| `LineString` | `MGLPolylineFeature` |
| `Polygon` | `MGLPolygonFeature` |
| `MultiPoint` | `MGLPointCollectionFeature` |
| `MultiLineString` | `MGLMultiPolylineFeature` |
| `MultiPolygon` | `MGLMultiPolygonFeature` |
| `GeometryCollection` | `MGLShapeCollectionFeature` |

Each of these MapboxGL types also receive convenience initializers that accept
the corresponding GEOSwift type.

## Contributing

To make a contribution:

* Fork the repo
* Start from the `main` branch and create a branch with a name that describes
  your contribution
* Follow the Mapbox Maps SDK [installation
  instructions](https://docs.mapbox.com/ios/maps/guides/install/) to configure
  your `~/.netrc` file.
* Run `$ pod install`
* Open GEOSwiftMapboxGL.xcworkspace and make your changes.
* Run `$ swiftlint` from the repo root and resolve any issues.
* Push your branch and create a pull request to `main`
* One of the maintainers will review your code and may request changes
* If your pull request is accepted, one of the maintainers should update the
  changelog before merging it.
* Due to the need for a secret Mapbox token to install the Mapbox SDK, CI will
  not run for PRs from forks. Maintainers should be sure to run the test suite
  locally before accepting any changes.

## Maintainer

* Andrew Hershberger ([@macdrevx](https://github.com/macdrevx))

## Past Maintainers

* Virgilio Favero Neto ([@vfn](https://github.com/vfn))
* Andrea Cremaschi ([@andreacremaschi](https://twitter.com/andreacremaschi))
  (original author)

## License

* GEOSwiftMapboxGL was released by Andrea Cremaschi
  ([@andreacremaschi](https://twitter.com/andreacremaschi)) under a MIT license.
  See LICENSE for more information.
* GEOSwift was released by Andrea Cremaschi
  ([@andreacremaschi](https://twitter.com/andreacremaschi)) under a MIT license.
  See LICENSE for more information.
* [GEOS](http://trac.osgeo.org/geos/) stands for Geometry Engine - Open Source,
  and is a C++ library, ported from the
  [Java Topology Suite](http://sourceforge.net/projects/jts-topo-suite/). GEOS
  implements the OpenGIS
  [Simple Features for SQL](http://www.opengeospatial.org/standards/sfs) spatial
  predicate functions and spatial operators. GEOS, now an OSGeo project, was
  initially developed and maintained by
  [Refractions Research](http://www.refractions.net/) of Victoria, Canada.


CHANGELOG.md
## 2.0.0

* [#11](https://github.com/GEOSwift/GEOSwiftMapboxGL/pull/11) Updated to
  GEOSwift 8, Swift 5, MapboxGL 6


GEOSwiftMapboxGL/GEOSwift+MapboxGL.swift
import CoreLocation
import Foundation
import GEOSwift
import Mapbox

public extension CLLocationCoordinate2D {
    init(_ point: Point) {
        self.init(latitude: point.y, longitude: point.x)
    }
}

public extension Point {
    init(longitude: Double, latitude: Double) {
        self.init(x: longitude, y: latitude)
    }

    init(_ coordinate: CLLocationCoordinate2D) {
        self.init(x: coordinate.longitude, y: coordinate.latitude)
    }
}

public extension GEOSwift.Polygon {
    // swiftlint:disable:next force_try
    static let world = try! GEOSwift.Polygon(
        exterior: Polygon.LinearRing(
            points: [
                Point(x: -180, y: 90),
                Point(x: -180, y: -90),
                Point(x: 180, y: -90),
                Point(x: 180, y: 90),
                Point(x: -180, y: 90)]))
}

public extension MGLPointAnnotation {
    convenience init(point: Point) {
        self.init()
        self.coordinate = CLLocationCoordinate2D(point)
    }
}

public extension MGLPolyline {
    convenience init(lineString: LineString) {
        var points = lineString.points.map(CLLocationCoordinate2D.init)
        self.init(coordinates: &points, count: UInt(points.count))
    }
}

public extension MGLPolygon {
    convenience init(polygon: GEOSwift.Polygon) {
        var exteriorCoordinates = polygon.exterior.points.map(CLLocationCoordinate2D.init)
        self.init(
            coordinates: &exteriorCoordinates,
            count: UInt(exteriorCoordinates.count),
            interiorPolygons: polygon.holes.map(MGLPolygon.init))
    }

    convenience init(linearRing: GEOSwift.Polygon.LinearRing) {
        var coordinates = linearRing.points.map(CLLocationCoordinate2D.init)
        self.init(coordinates: &coordinates, count: UInt(coordinates.count))
    }
}

public extension MGLPointCollection {
    convenience init(multiPoint: MultiPoint) {
        var coordinates = multiPoint.points.map(CLLocationCoordinate2D.init)
        self.init(coordinates: &coordinates, count: UInt(coordinates.count))
    }
}

public extension MGLMultiPolyline {
    convenience init(multiLineString: MultiLineString) {
        self.init(polylines: multiLineString.lineStrings.map(MGLPolyline.init))
    }
}

public extension MGLMultiPolygon {
    convenience init(multiPolygon: MultiPolygon) {
        self.init(polygons: multiPolygon.polygons.map(MGLPolygon.init))
    }
}

public extension MGLShapeCollection {
    convenience init(geometryCollection: GeometryCollection) {
        self.init(shapes: geometryCollection.geometries.map(MGLShape.make))
    }
}

public extension MGLShape {
    static func make(with geometry: GeometryConvertible) -> MGLShape {
        switch geometry.geometry {
        case let .point(point):
            return MGLPointAnnotation(point: point)
        case let .lineString(lineString):
            return MGLPolyline(lineString: lineString)
        case let .polygon(polygon):
            return MGLPolygon(polygon: polygon)
        case let .multiPoint(multiPoint):
            return MGLPointCollection(multiPoint: multiPoint)
        case let .multiLineString(multiLineString):
            return MGLMultiPolyline(multiLineString: multiLineString)
        case let .multiPolygon(multiPolygon):
            return MGLMultiPolygon(multiPolygon: multiPolygon)
        case let .geometryCollection(geometryCollection):
            return MGLShapeCollection(geometryCollection: geometryCollection)
        }
    }

    static func makeFeature(with geometry: GeometryConvertible) -> MGLShape & MGLFeature {
        switch geometry.geometry {
        case let .point(point):
            return MGLPointFeature(point: point)
        case let .lineString(lineString):
            return MGLPolylineFeature(lineString: lineString)
        case let .polygon(polygon):
            return MGLPolygonFeature(polygon: polygon)
        case let .multiPoint(multiPoint):
            return MGLPointCollectionFeature(multiPoint: multiPoint)
        case let .multiLineString(multiLineString):
            return MGLMultiPolylineFeature(multiLineString: multiLineString)
        case let .multiPolygon(multiPolygon):
            return MGLMultiPolygonFeature(multiPolygon: multiPolygon)
        case let .geometryCollection(geometryCollection):
            return MGLShapeCollectionFeature(geometryCollection: geometryCollection)
        }
    }
}


GEOSwiftMapboxGLTests/MapboxGLTests.swift
//  Copyright (c) 2017 GEOSwift. All rights reserved.

import Foundation
import XCTest

import GEOSwift
import Mapbox
@testable import GEOSwiftMapboxGL

final class MapboxGLTests: XCTestCase {

    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }

    func testCreateMKPointAnnotationFromPoint() {
        var result = false
        if let point = Geometry.create("POINT(45 9)") as? Waypoint,
            let _ = point.mapboxShape() as? MGLPointAnnotation {
                result = true
        }
        XCTAssert(result, "MGLPointAnnotation test failed")
    }

    func testCreateMKPolylineFromLineString() {
        var result = false
        let WKT = "LINESTRING(3 4,10 50,20 25)"
        if let linestring = Geometry.create(WKT) as? LineString,
            let _ = linestring.mapboxShape() as? MGLPolyline {
                result = true
        }
        XCTAssert(result, "MGLPolyline test failed")
    }

    func testCreateMKPolygonFromPolygon() {
        var result = false
        let WKT = "POLYGON((35 10, 45 45, 15 40, 10 20, 35 10),(20 30, 35 35, 30 20, 20 30))"
        if let polygon = Geometry.create(WKT) as? Polygon,
            let _ = polygon.mapboxShape() as? MGLPolygon {
                result = true
        }
        XCTAssert(result, "MGLPolygon test failed")
    }

    func testCreateMKShapesCollectionFromGeometryCollection() {
        var result = false
        let WKT = "GEOMETRYCOLLECTION(POINT(4 6),LINESTRING(4 6,7 10))"
        if let geometryCollection = Geometry.create(WKT) as? GeometryCollection,
            let _ = geometryCollection.mapboxShape() as? MGLShapesCollection {
                result = true
        }
        XCTAssert(result, "MGLShapesCollection test failed")
    }
}


GEOSwiftMapboxGL/MapboxGLAnnotations.swift
import CoreLocation
import Foundation
import GEOSwift
import Mapbox

public extension CLLocationCoordinate2D {
    init(_ point: Point) {
        self.init(latitude: point.y, longitude: point.x)
    }
}

public extension MGLPointAnnotation {
    convenience init(_ point: Point) {
        self.init()
        self.coordinate = CLLocationCoordinate2D(point)
    }
}

public extension MGLPolyline {
    convenience init(_ lineString: LineString) {
        var points = lineString.points.map(CLLocationCoordinate2D.init)
        self.init(coordinates: &points, count: UInt(points.count))
    }
}

public extension MGLPolygon {
    convenience init(_ polygon: Polygon) {
        var exteriorCoordinates = polygon.exterior.points.map(CLLocationCoordinate2D.init)
        self.init(
            coordinates: &exteriorCoordinates,
            count: UInt(exteriorCoordinates.count),
            interiorPolygons: polygon.holes.map(MGLPolygon.init))
    }

    convenience init(_ linearRing: Polygon.LinearRing) {
        var exteriorCoordinates = linearRing.points.map(CLLocationCoordinate2D.init)
        self.init(
            coordinates: &exteriorCoordinates,
            count: UInt(exteriorCoordinates.count))
    }
}

public extension MGLPointCollection {
    convenience init(_ multiPoint: MultiPoint) {
        var coordinates = multiPoint.points.map(CLLocationCoordinate2D.init)
        self.init(coordinates: &coordinates, count: UInt(coordinates.count))
    }
}

public extension MGLMultiPolyline {
    convenience init(_ multiLineString: MultiLineString) {
        self.init(polylines: multiLineString.lineStrings.map(MGLPolyline.init))
    }
}

public extension MGLMultiPolygon {
    convenience init(_ multiPolygon: MultiPolygon) {
        self.init(polygons: multiPolygon.polygons.map(MGLPolygon.init))
    }
}

public extension MGLShapeCollection {
    convenience init(_ geometryCollection: GeometryCollection) {
        self.init(shapes: geometryCollection.geometries.map(MGLShape.make))
    }
}

public extension MGLShape {

    static func make(with geometry: GeometryConvertible) -> MGLShape {
        switch geometry.geometry {
        case let .point(point):
            return MGLPointAnnotation(point)
        case let .lineString(lineString):
            return MGLPolyline(lineString)
        case let .polygon(polygon):
            return MGLPolygon(polygon)
        case let .multiPoint(multiPoint):
            return MGLPointCollection(multiPoint)
        case let .multiLineString(multiLineString):
            return MGLMultiPolyline(multiLineString)
        case let .multiPolygon(multiPolygon):
            return MGLMultiPolygon(multiPolygon)
        case let .geometryCollection(geometryCollection):
            return MGLShapeCollection(geometryCollection)
        }
    }

    static func makeFeature(with geometry: GeometryConvertible) -> MGLShape & MGLFeature {
        switch geometry.geometry {
        case let .point(point):
            return MGLPointFeature(point)
        case let .lineString(lineString):
            return MGLPolylineFeature(lineString)
        case let .polygon(polygon):
            return MGLPolygonFeature(polygon)
        case let .multiPoint(multiPoint):
            return MGLPointCollectionFeature(multiPoint)
        case let .multiLineString(multiLineString):
            return MGLMultiPolylineFeature(multiLineString)
        case let .multiPolygon(multiPolygon):
            return MGLMultiPolygonFeature(multiPolygon)
        case let .geometryCollection(geometryCollection):
            return MGLShapeCollectionFeature(geometryCollection)
        }
    }
}


GEOSwiftMapboxGLTests/GEOSwift+MapboxGLTests.swift
import GEOSwift
import GEOSwiftMapboxGL
import Mapbox
import XCTest

final class MapboxGLTests: XCTestCase {

    // CLLocationCoordinate2D

    func testCreateCLLocationCoordinate2DFromPoint() {
        let point = Point(x: 45, y: 9)

        let coordinate = CLLocationCoordinate2D(point)

        XCTAssertEqual(coordinate, CLLocationCoordinate2D(latitude: 9, longitude: 45))
    }

    // Point

    func testCreatePointWithLatAndLong() {
        let point = Point(longitude: 1, latitude: 2)

        XCTAssertEqual(point.x, 1)
        XCTAssertEqual(point.y, 2)
    }

    func testCreatePointWithCLLocationCoordinate2D() {
        let coord = CLLocationCoordinate2D(latitude: 2, longitude: 1)

        XCTAssertEqual(Point(coord), Point(x: 1, y: 2))
    }

    // Polygon

    func testWorldPolygon() {
        // just make sure it doesn't crash
        _ = GEOSwift.Polygon.world
    }

    // MGLPointAnnotation

    func verifyMGLPointCreation<T>(with factory: (Point) -> MGLShape,
                                   expectedType: T.Type,
                                   line: UInt = #line) where T: MGLPointAnnotation {
        let point = Point(x: 45, y: 9)

        let shape = factory(point)

        XCTAssertTrue(shape.isMember(of: T.self), line: line)
        XCTAssertEqual(shape.coordinate, CLLocationCoordinate2D(latitude: 9, longitude: 45), line: line)
    }

    func testCreateMGLPointAnnotationFromPoint() {
        verifyMGLPointCreation(with: MGLPointAnnotation.init(point:), expectedType: MGLPointAnnotation.self)
    }

    func testCreateMGLPointAnnotationFromPointViaMGLShapeMake() {
        verifyMGLPointCreation(with: MGLShape.make(with:), expectedType: MGLPointAnnotation.self)
    }

    func testCreateMGLPointFeatureFromPoint() {
        verifyMGLPointCreation(with: MGLPointFeature.init(point:), expectedType: MGLPointFeature.self)
    }

    func testCreateMGLPointFeatureFromPointViaMGLShapeMakeFeature() {
        verifyMGLPointCreation(with: MGLShape.makeFeature(with:), expectedType: MGLPointFeature.self)
    }

    // MGLPolyline

    func verifyMGLPolylineCreation<T>(with factory: (LineString) -> MGLShape,
                                      expectedType: T.Type,
                                      line: UInt = #line) where T: MGLPolyline {
        let lineString = try! LineString(wkt: "LINESTRING(3 4,10 50,20 25)")

        let shape = factory(lineString)

        XCTAssertTrue(shape.isMember(of: T.self), line: line)
        guard let polyline = shape as? T else {
            return
        }
        XCTAssertEqual(polyline.pointCount, 3, line: line)
    }

    func testCreateMGLPolylineFromLineString() {
        verifyMGLPolylineCreation(with: MGLPolyline.init(lineString:), expectedType: MGLPolyline.self)
    }

    func testCreateMGLPolylineFromLineStringViaMGLShapeMake() {
        verifyMGLPolylineCreation(with: MGLShape.make(with:), expectedType: MGLPolyline.self)
    }

    func testCreateMGLPolylineFeatureFromLineString() {
        verifyMGLPolylineCreation(
            with: MGLPolylineFeature.init(lineString:),
            expectedType: MGLPolylineFeature.self)
    }

    func testCreateMGLPolylineFeatureFromLineStringViaMGLShapeMakeFeature() {
        verifyMGLPolylineCreation(
            with: MGLShape.makeFeature(with:),
            expectedType: MGLPolylineFeature.self)
    }

    // MGLPolygon

    func testCreateMGLPolygonFromLinearRing() {
        let linearRing = try! Polygon.LinearRing(
            points: [
                Point(x: 35, y: 10),
                Point(x: 45, y: 45),
                Point(x: 15, y: 40),
                Point(x: 10, y: 20),
                Point(x: 35, y: 10)])

        let mglPolygon = MGLPolygon(linearRing: linearRing)

        XCTAssertEqual(mglPolygon.pointCount, 5)
        XCTAssertNil(mglPolygon.interiorPolygons)
    }

    func verifyMGLPolygonCreation<T>(with factory: (Polygon) -> MGLShape,
                                     expectedType: T.Type,
                                     line: UInt = #line) where T: MGLPolygon {
        let polygon = try! Polygon(
            wkt: "POLYGON((35 10, 45 45, 15 40, 10 20, 35 10),(20 30, 35 35, 30 20, 20 30))")

        let shape = factory(polygon)

        XCTAssertTrue(shape.isMember(of: T.self), line: line)
        guard let mglPolygon = shape as? T else {
            return
        }
        XCTAssertEqual(mglPolygon.pointCount, 5, line: line)
        XCTAssertEqual(mglPolygon.interiorPolygons?.count, 1, line: line)
        XCTAssertEqual(mglPolygon.interiorPolygons?.first?.pointCount, 4, line: line)
    }

    func testCreateMGLPolygonFromPolygon() {
        verifyMGLPolygonCreation(with: MGLPolygon.init(polygon:), expectedType: MGLPolygon.self)
    }

    func testCreateMGLPolygonFromPolygonViaMGLShapeMake() {
        verifyMGLPolygonCreation(with: MGLShape.make(with:), expectedType: MGLPolygon.self)
    }

    func testCreateMGLPolygonFeatureFromPolygon() {
        verifyMGLPolygonCreation(with: MGLPolygonFeature.init(polygon:), expectedType: MGLPolygonFeature.self)
    }

    func testCreateMGLPolygonFeatureFromPolygonViaMGLShapeMakeFeature() {
        verifyMGLPolygonCreation(with: MGLShape.makeFeature(with:), expectedType: MGLPolygonFeature.self)
    }

    // MGLPointCollection

    func verifyMGLPointCollectionCreation<T>(with factory: (MultiPoint) -> MGLShape,
                                             expectedType: T.Type,
                                             line: UInt = #line) where T: MGLPointCollection {
        let multiPoint = MultiPoint(points: [Point(x: 0, y: 1), Point(x: 2, y: 3)])

        let shape = factory(multiPoint)

        XCTAssertTrue(shape.isMember(of: T.self), line: line)
        guard let mglPointCollection = shape as? T else {
            return
        }
        XCTAssertEqual(mglPointCollection.pointCount, 2, line: line)
        var coordinate0 = CLLocationCoordinate2D(latitude: .nan, longitude: .nan)
        mglPointCollection.getCoordinates(&coordinate0, range: NSRange(location: 0, length: 1))
        var coordinate1 = CLLocationCoordinate2D(latitude: .nan, longitude: .nan)
        mglPointCollection.getCoordinates(&coordinate1, range: NSRange(location: 1, length: 1))
        XCTAssertEqual(coordinate0, CLLocationCoordinate2D(latitude: 1, longitude: 0), line: line)
        XCTAssertEqual(coordinate1, CLLocationCoordinate2D(latitude: 3, longitude: 2), line: line)
    }

    func testCreateMGLPointCollectionFromMultiPoint() {
        verifyMGLPointCollectionCreation(
            with: MGLPointCollection.init(multiPoint:),
            expectedType: MGLPointCollection.self)
    }

    func testCreateMGLPointCollectionFromMultiPointViaMGLShapeMake() {
        verifyMGLPointCollectionCreation(with: MGLShape.make(with:), expectedType: MGLPointCollection.self)
    }

    func testCreateMGLPointCollectionFeatureFromMultiPoint() {
        verifyMGLPointCollectionCreation(
            with: MGLPointCollectionFeature.init(multiPoint:),
            expectedType: MGLPointCollectionFeature.self)
    }

    func testCreateMGLPointCollectionFeatureFromMultiPointViaMGLShapeMakeFeature() {
        verifyMGLPointCollectionCreation(
            with: MGLShape.makeFeature(with:),
            expectedType: MGLPointCollectionFeature.self)
    }

    // MGLMultiPolyline

    func verifyMGLMultiPolylineCreation<T>(with factory: (MultiLineString) -> MGLShape,
                                           expectedType: T.Type,
                                           line: UInt = #line) where T: MGLMultiPolyline {
        let multiLineString = try! MultiLineString(lineStrings: [
            LineString(points: [Point(x: 0, y: 0), Point(x: 0, y: 1)]),
            LineString(points: [Point(x: 0, y: 0), Point(x: 1, y: 0)])])

        let shape = factory(multiLineString)

        XCTAssertTrue(shape.isMember(of: T.self), line: line)
        guard let mglMultiPolyline = shape as? T else {
            return
        }
        XCTAssertEqual(mglMultiPolyline.polylines.count, multiLineString.lineStrings.count, line: line)
    }

    func testCreateMGLMultiPolylineFromMultiLineString() {
        verifyMGLMultiPolylineCreation(
            with: MGLMultiPolyline.init(multiLineString:),
            expectedType: MGLMultiPolyline.self)
    }

    func testCreateMGLMultiPolylineFromMultiLineStringViaMGLShapeMake() {
        verifyMGLMultiPolylineCreation(with: MGLShape.make(with:), expectedType: MGLMultiPolyline.self)
    }

    func testCreateMGLMultiPolylineFeatureFromMultiLineString() {
        verifyMGLMultiPolylineCreation(
            with: MGLMultiPolylineFeature.init(multiLineString:),
            expectedType: MGLMultiPolylineFeature.self)
    }

    func testCreateMGLMultiPolylineFeatureFromMultiLineStringViaMGLShapeMakeFeature() {
        verifyMGLMultiPolylineCreation(
            with: MGLShape.makeFeature(with:),
            expectedType: MGLMultiPolylineFeature.self)
    }

    // MGLMultiPolygon

    func verifyMGLMultiPolygonCreation<T>(with factory: (MultiPolygon) -> MGLShape,
                                          expectedType: T.Type,
                                          line: UInt = #line) where T: MGLMultiPolygon {
        let multiPolygon = try! MultiPolygon(polygons: [
            Polygon(wkt: "POLYGON((35 10, 45 45, 15 40, 10 20, 35 10),(20 30, 35 35, 30 20, 20 30))"),
            Polygon(wkt: "POLYGON((35 10, 45 45, 15 40, 10 20, 35 10))")])

        let shape = factory(multiPolygon)

        XCTAssertTrue(shape.isMember(of: T.self), line: line)
        guard let mglMultiPolygon = shape as? T else {
            return
        }
        XCTAssertEqual(mglMultiPolygon.polygons.count, multiPolygon.polygons.count, line: line)
    }

    func testCreateMGLMultiPolygonFromMultiPolygon() {
        verifyMGLMultiPolygonCreation(
            with: MGLMultiPolygon.init(multiPolygon:),
            expectedType: MGLMultiPolygon.self)
    }

    func testCreateMGLMultiPolygonFromMultiPolygonViaMGLShapeMake() {
        verifyMGLMultiPolygonCreation(
            with: MGLShape.make(with:),
            expectedType: MGLMultiPolygon.self)
    }

    func testCreateMGLMultiPolygonFeatureFromMultiPolygon() {
        verifyMGLMultiPolygonCreation(
            with: MGLMultiPolygonFeature.init(multiPolygon:),
            expectedType: MGLMultiPolygonFeature.self)
    }

    func testCreateMGLMultiPolygonFeatureFromMultiPolygonViaMGLShapeMakeFeature() {
        verifyMGLMultiPolygonCreation(
            with: MGLShape.makeFeature(with:),
            expectedType: MGLMultiPolygonFeature.self)
    }

    // MGLShapeCollection

    func verifyMGLShapeCollectionCreation<T>(with factory: (GeometryCollection) -> MGLShape,
                                             expectedType: T.Type,
                                             line: UInt = #line) where T: MGLShapeCollection {
        let geometryCollection = GeometryCollection(geometries: [Point(x: 0, y: 1)])

        let shape = factory(geometryCollection)

        XCTAssertTrue(shape.isMember(of: T.self))
        guard let shapeCollection = shape as? T else {
            return
        }
        XCTAssertEqual(shapeCollection.shapes.count, 1, line: line)
        XCTAssertTrue(shapeCollection.shapes.first is MGLPointAnnotation, line: line)
        XCTAssertEqual(
            shapeCollection.shapes.first?.coordinate,
            CLLocationCoordinate2D(latitude: 1, longitude: 0),
            line: line)
    }

    func testCreateMGLShapeCollectionFromGeometryCollection() {
        verifyMGLShapeCollectionCreation(
            with: MGLShapeCollection.init(geometryCollection:),
            expectedType: MGLShapeCollection.self)
    }

    func testCreateMGLShapeCollectionFromGeometryCollectionViaMGLShapeMake() {
        verifyMGLShapeCollectionCreation(with: MGLShape.make(with:), expectedType: MGLShapeCollection.self)
    }

    func testCreateMGLShapeCollectionFeatureFromGeometryCollection() {
        verifyMGLShapeCollectionCreation(
            with: MGLShapeCollectionFeature.init(geometryCollection:),
            expectedType: MGLShapeCollectionFeature.self)
    }

    func testCreateMGLShapeCollectionFeatureFromGeometryCollectionViaMGLShapeMakeFeature() {
        verifyMGLShapeCollectionCreation(
            with: MGLShape.makeFeature(with:),
            expectedType: MGLShapeCollectionFeature.self)
    }
}
